src/site_upload/check_if_complete/check_if_complete.py

```diff
@@ -0,0 +1,140 @@
+            # processed is a minute, and since this is largely parallel it shouldn't
+            # change by much; we'll set a large buffer just in case, which should
+            # still exclude dangling files from error states
+            if upload_time > head["LastModified"] or (delta.seconds >= 300 and delta.seconds < 0):
+                return False, None
+    return True, transaction
+
+
+def has_new_packages(message, transaction) -> bool:
+    db = os.environ.get("GLUE_DB_NAME")
+    s3_bucket_name = os.environ.get("BUCKET_NAME")
+    tables = awswrangler.athena.read_sql_query(
+        (
+            f"SELECT table_name FROM information_schema.tables "  # noqa: S608
+            f"WHERE table_schema = '{db}' "  # nosec
+            f"AND regexp_like(table_name, '^{message['study']}__')"
+        ),
+        database=db,
+        s3_output=f"s3://{s3_bucket_name}/awswrangler",
+        workgroup=os.environ.get("WORKGROUP_NAME"),
+    )["table_name"].values.tolist()
+    for dp_type in ["cube", "annotated_cube", "flat"]:
+        filenames = transaction.get(dp_type, [])
+
+        for filename in filenames:
+            dp = filename.split("__")[1].split(".")[0]
+            if dp_type == "flat":
+                name = f"{message['study']}__{dp}__{message['site']}__{transaction['version']}"
+            else:
+                name = f"{message['study']}__{dp}__{transaction['version']}"
+            if name not in tables:
+                return True
+    return False
+
+
+def cleanup_transaction(message) -> bool:
+    key = f"{enums.BucketPath.META.value}/transactions/{message['site']}__{message['study']}.json"
+    try:
+        s3_client.head_object(Bucket=os.environ.get("BUCKET_NAME"), Key=key)
+    except botocore.exceptions.ClientError:
+        return False
+    s3_client.delete_object(Bucket=os.environ.get("BUCKET_NAME"), Key=key)
+    return True
+
+
+# @decorators.generic_error_handler(msg="Error processing metadata events")
+def check_if_complete_handler(event, context):
+    del context
+    message = json.loads(event["Records"][0]["Sns"]["Message"])
+    completed, transaction = check_if_complete(message)
+    if not completed:
+        return functions.http_response(200, "Processing not completed")
+    new = has_new_packages(message, transaction)
+
+    if new:
+        attempts = 0
+        while attempts < 10:
+            crawler = g_client.get_crawler(Name=os.environ.get("GLUE_CRAWLER_NAME"))["Crawler"]
+            if crawler["State"] == "READY":
+                cleanup_performed = cleanup_transaction(message)
+                if not cleanup_performed:
+                    return functions.http_response(
+                        200, f"Processing request for {message!s} already sent"
+                    )
+                if "LastCrawl" in crawler and crawler["LastCrawl"][
+                    "StartTime"
+                ] > datetime.datetime.fromisoformat(event["Records"][0]["Sns"]["Timestamp"]):
```

Comment:
This got formatted real weird by ruff - maybe use some variables to keep the line length down.